### ISSUE TITLE
Unique wire by id?

### ISSUE THREAD

**Issue by uhop** [WebReflection/hyperHTML#93]
hyperHTML provides a way to "cache" wire using an object (WeakMap?). The most common case to get objects are from a server using some transport envelop, e.g., JSON, which creates new objects every single time, when we unpack it. It makes the whole mechanism largely useless.

For example I got a page of 10 objects from the server, rendered them, then a user decided to sort them on a server and retrieving them again. Some of objects in a coming bundle are the same, some are different. But all of them will be considered new objects in a WeakMap sense. I would prefer to use some unique features instead, like ids. Unfortunately them being primitives, they don't play well with WeakMap. Yes, I can create a layer to work around that, but wouldn't be more appropriate to add it directly to hyperHTML, e.g., as a simple map?

**Comment by WebReflection** [comment]
> It makes the whole mechanism largely useless.

It depeneds. To start with, not every web app load JSON from the backend at 60FPS, so that even if you trash every time the content, being the async update a one-off operation it shouldn't be an issue.

On top of that, efficient Web apps don't reload everything every time, they load an initial list of items and eventually load only changes, updates, on top of this item.

If you load every single time 1000 items you already have, I'd say the problem is not strictly the `wire` one.

> then a user decided to sort them on a server and retrieving them again

I don't know what this means. Sorted on a server? Just sort on the client and eventually update the sort order as single field on the server instead. Can you please provide a concrete example of what you mean here?

Again, in my mind this is same list to be sorted differently. If you need the server for that, I see only one reason: the list was of thousands of items and you show only 10.

In this case you would trash first ten when you want to show last 10 from position 990, right?

> I would prefer to use some unique features instead, like ids.

And that's indeed what I would use, and as a matter of fact, I've done that already indeed, here an example: https://github.com/WebReflection/hyperHTML/blob/master/test/dbmonster.html#L12-L15

What's that? It's a deadly simple workaround for JSON items with IDs.

```js
const wires = Object.create(null);
const renderID = id => wires[id] || (wires[id] = hyper());

fetch(info).then(b => b.json()).then(data => {
  hyper(document.body)`
  <ul>${data.items.map(item => renderID(item.id)`
  <li data-id=${item.id}>${item.name}</li>`)}</ul>`;
});
```

> wouldn't be more appropriate to add it directly to hyperHTML, e.g., as a simple map?

Imagine you have two lists of items, and you are dragging them around, and you use just IDs or names ... and you'll never have twice the same item in both lists because you file a bug hyperHTML doesn't understand primitives as IDs ... 

I don't want to deal with the following:

  * memory leaks
  * same primitive for two different objects

### and moreover ...

What you are proposing is already possible with `hyperHTML.wire`, it's the ID that you want, right?

```js
hyperHTML.wire(document, ':' + item.id)`<li>content</li>`;
```

And that's literally it. You have your primitive as id, and you relate it with an object that will never weakly disappear.

TL;DR if `wire(obj, :id)` doesn't work already, I'm afraid there's nothing else that could possibly work here.

**Comment by uhop** [comment]
> It depeneds. To start with, not every web app load JSON from the backend at 60FPS, so that even if you trash every time the content, being the async update a one-off operation it shouldn't be an issue.

It is not an issue.

Just out of curiosity: why do you want to rerender the same object 60 times per second with the same template? Clearly this is something I don't plan to do. I think I misunderstood the purpose of wire with id. 

> I don't know what this means. Sorted on a server? Just sort on the client and eventually update the sort order as single field on the server instead. Can you please provide a concrete example of what you mean here?

Most of programming work is in IT, and IT is mostly about CRUD: forms + tables. Tables frequently operates with million of records, but show relatively small amount of data a time (pagination). Typical operation on them (besides paging) include different sorting, and filtering by some criteria. Ultimately all this stuff is done by a database (on a server farm).

You are welcome to do it on the client, if you are ready to download database tables, which require a lot of bandwidth, and memory on a client. From my practice, it is prohibitive as a general solution.

I hope my example was concrete enough.

> What's that? It's a deadly simple workaround for JSON items with IDs.

Yes, I can do a dead simple workaround. I will even put it in my library. I bothered to alert you about it, because it looks like a general enough functionality.

(I understand now your issue about the "issue". I issued an issue because it appears to be the only way to file anything --- from a question, to an enhancement request, not necessarily a bug report. Most people use labels for that. I tried, but it turned out I cannot label an issue in any way. Sorry about that.)

> You have your primitive as id, and you relate it with an object that will never weakly disappear.

Thank you very much. From the documentation I understood that the second optional argument of `wire()` can be a type (specifically `"html"` vs. `"svg"`), because it's important to distinguish between html nodes, the assumed default, and svg nodes, which require special treatment during their creation, and ids, which are compatible with types, and should be used, if I want to create the layout once per object, reusing as much as possible so that object has a weak reference to its layouts. I failed to look how they actually used so didn't divine the trick of using a permanent object with different ids. So sorry.

What I do is very simple: I render a relatively small table from a bigger data source. When certain criteria are changed, it can be refreshed on the server, yet some objects will be the same, just with a different position. I wanted to save on rendering a row (just rearranging them would be enough), but why bother. It is not like I am going for a smooth animation, or other advanced things.

Keep up the good work!

**Comment by WebReflection** [comment]
> What I do is very simple

I understood that, and I know documentation is not the best one out there, but it's improving and the ID bit should be kinda self explanatory now.
https://viperhtml.js.org/hyperhtml/documentation/#api-1-2

In any case, you gave me an idea: use a global, never weak reference, automatically when a wire has a primitive as an ID.

`hyper.wire(primitive[, type])` could translate into `hyper.wire(window, (type || 'html') + ':' + primitive)` so that the signature/API outsmarts cases like yours.

I think it's a no-brainer, it just needs a lot of documentation to explain possible side effects such memory consumption, never cleaned, and possible collisions with the regular signature ...

but hey, everyone might be happy after all, isn't it?

**Comment by WebReflection** [comment]
You know what? You convinced me this is going to be a common use/case requirement, hence my PR against this use case.
https://github.com/WebReflection/hyperHTML/pull/95

It's backward compatible, it doesn't penalize performances much, it works for devs with similar use case.

Thanks.

**Comment by WebReflection** [comment]
Never mind, the night brought more thoughts and I've decided to drop the PR.
https://github.com/WebReflection/hyperHTML/pull/95#issuecomment-326195363

I will improve the documentation on this though.

Best Regards

**Comment by uhop** [comment]
At least add a documented way to manually clear the cache (WeakMap), because otherwise it can be a problem for long-running applications.

**Comment by WebReflection** [comment]
that's not going to happen because it's exactly why there is a WeakMap.

If you hold all nodes in memory you have exact same issue. If you trash them, they will clean the WeakMap.

WeakMap exists so that you don't care about cleaning it.

If you can clean a weak map, you are holding the value, hence you're not freeing anything.

So, speculations about non-existent real-world issues seems to be a bit off topic at this stage.

When you'll have that problem and you can show me the real-world use case, I'll eventually open a bug and fix it.

Meanwhile, nobody ever deleted a key from a WeakMap to my memories.

**Comment by WebReflection** [comment]
P.S. if you want to do that yourself, hold an empty object reference for as much as you want, and just `obj = null` when you're tired.

```js
var myWeakHolder = {};

for (var i = 0; i < allTheIDsOfTheUniverse.length; i++) {
  wire(myWeakHolder, ':' + allTheIDsOfTheUniverse[i]);
}

setTimeout(() => myWeakHolder = null, 5000);
```

**Comment by WebReflection** [comment]
P.S.2 if it's above example you were referring, then yes, it will be there in the docs.
